llvm/include/llvm/ABI/Types.h

```diff
@@ -0,0 +1,543 @@
+
+public:
+  StructType(ArrayRef<FieldInfo> StructFields, ArrayRef<FieldInfo> Bases,
+             ArrayRef<FieldInfo> VBases, TypeSize Size, Align Align,
+             StructPacking Pack = StructPacking::Default, bool Union = false,
+             bool CXXRecord = false, bool Polymorphic = false,
+             bool NonTrivialCopy = false, bool NonTrivialDtor = false,
+             bool FlexibleArray = false, bool UnalignedFields = false,
+             bool CanPassInRegs = false, bool IsCoercedStr = false,
+             bool Transparent = false)
+      : Type(TypeKind::Struct, Size, Align), Fields(StructFields),
+        BaseClasses(Bases), VirtualBaseClasses(VBases), Packing(Pack),
+        CanPassInRegisters(CanPassInRegs), IsCoercedStruct(IsCoercedStr),
+        IsUnion(Union), IsTransparent(Transparent), IsCXXRecord(CXXRecord),
+        IsPolymorphic(Polymorphic),
+        HasNonTrivialCopyConstructor(NonTrivialCopy),
+        HasNonTrivialDestructor(NonTrivialDtor),
+        HasFlexibleArrayMember(FlexibleArray),
+        HasUnalignedFields(UnalignedFields) {}
+
+  uint32_t getNumFields() const { return Fields.size(); }
+  StructPacking getPacking() const { return Packing; }
+
+  bool isUnion() const { return IsUnion; }
+  bool isCXXRecord() const { return IsCXXRecord; }
+  bool isPolymorphic() const { return IsPolymorphic; }
+  bool hasNonTrivialCopyConstructor() const {
+    return HasNonTrivialCopyConstructor;
+  }
+  bool isCoercedStruct() const { return IsCoercedStruct; }
+  bool canPassInRegisters() const { return CanPassInRegisters; }
+  bool hasNonTrivialDestructor() const { return HasNonTrivialDestructor; }
+  bool hasFlexibleArrayMember() const { return HasFlexibleArrayMember; }
+  bool hasUnalignedFields() const { return HasUnalignedFields; }
+
+  uint32_t getNumBaseClasses() const { return BaseClasses.size(); }
+  uint32_t getNumVirtualBaseClasses() const {
+    return VirtualBaseClasses.size();
+  }
+  bool isTransparentUnion() const { return IsTransparent; }
+  ArrayRef<FieldInfo> getFields() const { return Fields; }
+  ArrayRef<FieldInfo> getBaseClasses() const { return BaseClasses; }
+  ArrayRef<FieldInfo> getVirtualBaseClasses() const {
+    return VirtualBaseClasses;
+  }
+
+  bool isEmptyForABI(const llvm::abi::Type *Ty) const {
+    const auto *ST = dyn_cast<StructType>(Ty);
+    if (!ST)
+      return false;
+
+    for (const auto &Field : ST->getFields()) {
+      if (!Field.IsUnnamedBitfield)
+        return false;
+    }
+
+    if (ST->isCXXRecord()) {
+      for (const auto &Base : ST->getBaseClasses()) {
+        if (!isEmptyForABI(Base.FieldType))
+          return false;
+      }
+
+      for (const auto &VBase : ST->getVirtualBaseClasses()) {
+        if (!isEmptyForABI(VBase.FieldType))
+          return false;
+      }
+    }
+
+    return true;
+  }
+
+  const FieldInfo *getElementContainingOffset(unsigned OffsetInBits) const {
+    std::vector<std::pair<unsigned, const FieldInfo *>> AllElements;
```

Comment:
```suggestion
    SmallVector<std::pair<unsigned, const FieldInfo *>> AllElements;
```

Though I kind of wonder whether we shouldn't require the fields to already be in sorted order. Do they currently get inserted out of order?